src/custom/constants/networks.ts

```diff
@@ -1,30 +1,146 @@
+  [SupportedChainId.MAINNET]: [InfuraProvider, CloudflareProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GOERLI]: [InfuraProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GNOSIS_CHAIN]: [],
+}
+
+function getProvider(chainId: SupportedChainId): JsonRpcProvider[] {
+  const result: JsonRpcProvider[] = []
+
+  if (window.ethereum) {
+    result.push(new Web3Provider(window.ethereum, RPC_CONFIG[chainId]))
+  }
+
+  if (typeof RPC_CONFIG[chainId].rpcUrl === 'string') {
+    result.push(new StaticJsonRpcProvider(RPC_CONFIG[chainId].rpcUrl))
+  } else {
+    result.push(new InfuraProvider(RPC_CONFIG[chainId], INFURA_KEY))
+  }
+
+  if (result.length > 0) {
+    return result
+  }
+
+  for (const Provider of PUBLIC_PROVIDER_LIST[chainId]) {
+    try {
+      const provider = new Provider(RPC_CONFIG[chainId])
+
+      result.push(provider)
+    } catch (error) {
+      const { sentryError, tags } = constructSentryError(error, { message: "Couldn't create public provider" })
+
+      Sentry.captureException(sentryError, { tags })
+    }
+  }
+
+  return result
+}
+
+export const PROVIDERS: Record<SupportedChainId, JsonRpcProvider[]> = {
+  [SupportedChainId.MAINNET]: getProvider(SupportedChainId.MAINNET),
+  [SupportedChainId.GOERLI]: getProvider(SupportedChainId.GOERLI),
+  [SupportedChainId.GNOSIS_CHAIN]: getProvider(SupportedChainId.GNOSIS_CHAIN),
+}
+
+export const getRpcUrls = (chainId: SupportedChainId) => PROVIDERS[chainId].map((provider) => provider.connection.url)
+
+export const PROVIDERS_RPC_URLS: Record<SupportedChainId, string[]> = {
+  [SupportedChainId.MAINNET]: getRpcUrls(SupportedChainId.MAINNET),
+  [SupportedChainId.GOERLI]: getRpcUrls(SupportedChainId.GOERLI),
+  [SupportedChainId.GNOSIS_CHAIN]: getRpcUrls(SupportedChainId.GNOSIS_CHAIN),
+}
+
+export const PROVIDERS_RPC_URLS_FIRST_ONLY: Record<SupportedChainId, string> = {
+  [SupportedChainId.MAINNET]: PROVIDERS_RPC_URLS[SupportedChainId.MAINNET][0],
+  [SupportedChainId.GOERLI]: PROVIDERS_RPC_URLS[SupportedChainId.GOERLI][0],
+  [SupportedChainId.GNOSIS_CHAIN]: PROVIDERS_RPC_URLS[SupportedChainId.GNOSIS_CHAIN][0],
+}
+
+export const providers = {
+  get mainnet() {
+    const _providers = getProvider(SupportedChainId.MAINNET)
+
+    if (_providers.length > 0) {
+      return _providers[0]
+    } else {
+      const message = 'No provider found for MAINNET_PROVIDER'
+      const { sentryError, tags } = constructSentryError(new Error(), {
+        message,
+      })
+
+      Sentry.captureException(sentryError, { tags })
+
+      throw new Error(message)
+    }
+  },
+}
+
+function constructSentryError(baseError: unknown, { message }: { message: string }) {
+  const constructedError = Object.assign(new Error(), baseError, {
+    message,
+    name: 'ProviderError',
+  })
+
+  const tags = {
+    errorType: 'provider',
+  }
+
+  return { baseError, sentryError: constructedError, tags }
```

Comment:
why returning also the baseError that is not modified and part of the params?

Reply:
Removed it, good point.